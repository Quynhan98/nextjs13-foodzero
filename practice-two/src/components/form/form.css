.book-form {
  padding-top: 30px;
}

.group-btn {
  display: flex;
  margin-top: 20px;
  column-gap: 20px;
  justify-content: center;
}

.form-heading {
  font-size: var(--size-l);
  font-weight: bold;
  text-align: center;
  color: var(--color-primary);
}
