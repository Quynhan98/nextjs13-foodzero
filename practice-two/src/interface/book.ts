export interface Book {
  id: number;
  title: string;
  author: string;
  price: number;
  desc: string;
  image: string;
}
