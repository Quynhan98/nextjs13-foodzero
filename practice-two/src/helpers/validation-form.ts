import { VALIDATE_MESSAGES } from '@constant/messages';
import { Book } from '@interface/book';

export interface ValidationResult {
  isValid: boolean;
  errors?: {
    title: string;
    author: string;
    price: string;
    desc: string;
    img: string;
  }
}

/**
 * Validate empty input value
 * @param value input value
 * @returns true/false
 */
const isEmpty = (value: string | number): boolean => {
  return !!value;
};

/**
 * validate for form
 * @param book input value(Book)
 * @returns ValidationResult
 */
export const formValidate = (book: Book) => {
  const result: ValidationResult = { isValid: true };

  result.errors = {
    title: '',
    author: '',
    price: '',
    desc: '',
    img: ''
  };

  // Title
  if (!isEmpty(book.title)) {
    result.errors.title = VALIDATE_MESSAGES.BOOK_TITLE_REQUIRED;
  }

  // Author
  if (!isEmpty(book.author)) {
    result.errors.author = VALIDATE_MESSAGES.AUTHOR_REQUIRED;
  }

  // Price
  if (!isEmpty(book.price)) {
    result.errors.price = VALIDATE_MESSAGES.PRICE_REQUIRED;
  }

  if (isNaN(book.price) || book.price < 0) {
    result.errors.price = VALIDATE_MESSAGES.PRICE_FORMAT;
  }

  // Desc
  if (!isEmpty(book.desc)) {
    result.errors.desc = VALIDATE_MESSAGES.DESC_REQUIRED;
  }

  // Image
  if (!isEmpty(book.image)) {
    result.errors.img = VALIDATE_MESSAGES.IMAGE_REQUIRED;
  }

  // Result
  const validate = result.errors;

  if (validate.author || validate.desc || validate.img || validate.price || validate.title) {
    result.isValid = false;
  }

  return result;
};
