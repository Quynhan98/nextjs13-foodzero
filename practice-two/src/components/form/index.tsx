import { useState } from 'react';
import { Book } from '@interface/book';
import { formValidate, ValidationResult } from '@helpers/validation-form';
import { Button } from '@components/common/button';
import { Textbox } from '@components/common/textbox';
import { Uploader } from '@components/common/uploader';
import { Textarea } from '@components/common/textarea';
import './form.css';

export type FormProp = {
  selectedBook?: Book;
  onHandleClose?: () => void;
  onCreate: (book: Book) => void;
  onHandleEdit: (book: Book) => void;
};

export type InputProp = {
  title?: string;
  author?: string;
  price?: number;
  desc?: string;
  image?: string;
  id?: number;
};

export type MessagesErr = {
  title: string;
  author: string;
  price: string;
  desc: string;
  img: string;
};

const initialBook: Book = {
  id: 0,
  title: '',
  author: '',
  price: 0,
  desc: '',
  image: '',
};

const initialErrorMsgs: MessagesErr = {
  title: '',
  author: '',
  price: '',
  desc: '',
  img: '',
};

export const Form = (props: FormProp): JSX.Element => {
  const { selectedBook, onHandleClose, onCreate, onHandleEdit } = props;
  const [book, setBook] = useState<Book>(selectedBook || initialBook);
  const [msgError, setMsgError] = useState(initialErrorMsgs);

  const uploadImage = (img: string) => {
    setBook({ ...book, image: img });
  };

  const handleInputValue = (value: InputProp) => {
    setBook({ ...book, ...value });
  };

  const handleBlur = (): void => {
    const validateForm: ValidationResult = formValidate(book);

    if (!validateForm.isValid) {
      const messages = validateForm.errors!;

      setMsgError(messages);
    } else {
      setMsgError({} as MessagesErr);
    }
  };

  const handleCreate = (): void => {
    const newBook: Book = { ...book, id: Date.now() };

    onCreate(newBook);
  };

  const handleEdit = (): void => {
    onHandleEdit(book);
  };

  return (
    <div className="container">
      <h2 className="form-heading">{selectedBook ? 'Edit Book' : 'Create New Book'}</h2>
      <form data-testid="form" className="book-form">
        <Textbox
          onHandleBlur={handleBlur}
          handleInput={handleInputValue}
          label="Title"
          inputType="text"
          name="title"
          value={book.title}
          messageErr={msgError.title}
        />
        <Textbox
          onHandleBlur={handleBlur}
          handleInput={handleInputValue}
          label="Author"
          inputType="text"
          name="author"
          value={book.author}
          messageErr={msgError.author}
        />
        <Textbox
          onHandleBlur={handleBlur}
          handleInput={handleInputValue}
          label="Price"
          inputType="text"
          name="price"
          value={book.price}
          messageErr={msgError.price}
        />
        <Textarea
          onHandleBlur={handleBlur}
          rows={3}
          handleInput={handleInputValue}
          label="Description"
          name="desc"
          value={book.desc}
          messageErr={msgError.desc}
        />
        <Uploader
          onHandleBlur={handleBlur}
          handleChange={uploadImage}
          baseImage={book.image}
          label="Image"
          name="image"
          messageErr={msgError.img}
        />
        <div className="group-btn">
          <Button
            handleClick={selectedBook ? handleEdit : handleCreate}
            typeButton="button"
            size="btn-small"
            color="btn-secondary"
          >
            {selectedBook ? 'Update' : 'Create'}
          </Button>
          <Button
            handleClick={onHandleClose}
            typeButton="button"
            size="btn-small"
            color="btn-warning"
          >
            Cancel
          </Button>
        </div>
      </form>
    </div>
  );
};
