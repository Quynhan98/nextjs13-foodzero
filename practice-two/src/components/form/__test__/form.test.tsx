import { fireEvent, render, screen } from '@testing-library/react';
import { Form, FormProp, MessagesErr } from '@components/form';
import { formValidate } from '@helpers/validation-form';
import { Book } from '@interface/book';

describe('Test Form Component', () => {
  const initialBook: Book = {
    id: 0,
    title: 'Book 1',
    author: 'Author 1',
    price: 1,
    desc: 'Description',
    image: 'base_img',
  };

  const initialValue: FormProp = {
    onHandleClose: jest.fn(),
    onCreate: jest.fn(),
    onHandleEdit: jest.fn(),
  };

  const initialFormSelected: FormProp = {
    selectedBook: initialBook,
    onHandleClose: jest.fn(),
    onCreate: jest.fn(),
    onHandleEdit: jest.fn(),
  };

  test('Component Form matchers DOM Snapshot', () => {
    const component = render(<Form {...initialValue} />);

    expect(component).toMatchSnapshot();
  });

  test('Component Form calls handleChange function when user click', () => {
    render(<Form {...initialValue} />);

    fireEvent.click(screen.getByText('Create'));
    expect(initialValue.onCreate).toBeCalled();
  });

  test('Component Form handle Input when user enter value', () => {
    render(<Form {...initialValue} />);

    fireEvent.change(screen.getAllByRole('textbox')[0], { target: { value: 'Nhan' } });
    expect(screen.getAllByRole('textbox')[0]).toHaveValue('Nhan');
  });

  test('Component Form handle blur', () => {
    render(<Form {...initialValue} />);

    const mockBookData: Book = {
      id: 1,
      title: 'title',
      author: 'nhan',
      price: 123,
      desc: 'this is a book',
      image: 'url',
    };

    fireEvent.blur(screen.getAllByRole('textbox')[0]);
    expect(formValidate(mockBookData).isValid).toEqual(true);
    expect(formValidate(mockBookData).isValid);
    expect(screen.getAllByRole('textbox')[0]).not.toHaveFocus();
  });

  test('Component Form handle blur with error message', () => {
    render(<Form {...initialValue} />);

    const mockBookData: Book = {
      id: 1,
      title: '',
      author: 'nhan',
      price: 123,
      desc: 'this is a book',
      image: 'url',
    };

    fireEvent.blur(screen.getAllByRole('textbox')[0]);
    expect(formValidate(mockBookData).isValid).toEqual(false);
    expect(screen.getAllByRole('textbox')[0]).not.toHaveFocus();
  });

  test('Component Form has been selected matchers DOM Snapshot', () => {
    const component = render(<Form {...initialFormSelected} />);

    expect(component).toMatchSnapshot();
  });

  test('Component Form has been selected calls handleChange function when user click', () => {
    render(<Form {...initialFormSelected} />);

    fireEvent.click(screen.getByText('Update'));
    expect(initialFormSelected.onHandleEdit).toBeCalled();
    expect(screen.getAllByRole('textbox')[0]).toHaveValue('Book 1');
  });
});
