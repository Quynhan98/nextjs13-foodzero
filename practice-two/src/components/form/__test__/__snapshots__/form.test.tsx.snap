// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Form Component Component Form has been selected matchers DOM Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <h2
          class="form-heading"
        >
          Edit Book
        </h2>
        <form
          class="book-form"
          data-testid="form"
        >
          <div
            class="input"
          >
            <label
              class="input-title"
            >
              Title
              <input
                class="input-content"
                name="title"
                type="text"
                value="Book 1"
              />
            </label>
          </div>
          <div
            class="input"
          >
            <label
              class="input-title"
            >
              Author
              <input
                class="input-content"
                name="author"
                type="text"
                value="Author 1"
              />
            </label>
          </div>
          <div
            class="input"
          >
            <label
              class="input-title"
            >
              Price
              <input
                class="input-content"
                name="price"
                type="text"
                value="1"
              />
            </label>
          </div>
          <div
            class="input"
          >
            <label
              class="input-title"
            >
              Description
              <textarea
                class="textarea-content"
                id="desc"
                name="desc"
                rows="3"
              >
                Description
              </textarea>
            </label>
          </div>
          <div
            class="uploader"
          >
            <label
              class="uploader-title"
            >
              Image
              <div
                class="uploader-input"
              >
                <input
                  class="uploader-content"
                  data-testid="uploader-img"
                  name="image"
                  type="file"
                />
                <img
                  class="uploader-image"
                  data-testid="image-input"
                  src="base_img"
                />
              </div>
            </label>
          </div>
          <div
            class="group-btn"
          >
            <button
              class="button btn-secondary btn-small"
              type="button"
            >
              Update
            </button>
            <button
              class="button btn-warning btn-small"
              type="button"
            >
              Cancel
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <h2
        class="form-heading"
      >
        Edit Book
      </h2>
      <form
        class="book-form"
        data-testid="form"
      >
        <div
          class="input"
        >
          <label
            class="input-title"
          >
            Title
            <input
              class="input-content"
              name="title"
              type="text"
              value="Book 1"
            />
          </label>
        </div>
        <div
          class="input"
        >
          <label
            class="input-title"
          >
            Author
            <input
              class="input-content"
              name="author"
              type="text"
              value="Author 1"
            />
          </label>
        </div>
        <div
          class="input"
        >
          <label
            class="input-title"
          >
            Price
            <input
              class="input-content"
              name="price"
              type="text"
              value="1"
            />
          </label>
        </div>
        <div
          class="input"
        >
          <label
            class="input-title"
          >
            Description
            <textarea
              class="textarea-content"
              id="desc"
              name="desc"
              rows="3"
            >
              Description
            </textarea>
          </label>
        </div>
        <div
          class="uploader"
        >
          <label
            class="uploader-title"
          >
            Image
            <div
              class="uploader-input"
            >
              <input
                class="uploader-content"
                data-testid="uploader-img"
                name="image"
                type="file"
              />
              <img
                class="uploader-image"
                data-testid="image-input"
                src="base_img"
              />
            </div>
          </label>
        </div>
        <div
          class="group-btn"
        >
          <button
            class="button btn-secondary btn-small"
            type="button"
          >
            Update
          </button>
          <button
            class="button btn-warning btn-small"
            type="button"
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test Form Component Component Form matchers DOM Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <h2
          class="form-heading"
        >
          Create New Book
        </h2>
        <form
          class="book-form"
          data-testid="form"
        >
          <div
            class="input"
          >
            <label
              class="input-title"
            >
              Title
              <input
                class="input-content"
                name="title"
                type="text"
                value=""
              />
            </label>
          </div>
          <div
            class="input"
          >
            <label
              class="input-title"
            >
              Author
              <input
                class="input-content"
                name="author"
                type="text"
                value=""
              />
            </label>
          </div>
          <div
            class="input"
          >
            <label
              class="input-title"
            >
              Price
              <input
                class="input-content"
                name="price"
                type="text"
                value="0"
              />
            </label>
          </div>
          <div
            class="input"
          >
            <label
              class="input-title"
            >
              Description
              <textarea
                class="textarea-content"
                id="desc"
                name="desc"
                rows="3"
              />
            </label>
          </div>
          <div
            class="uploader"
          >
            <label
              class="uploader-title"
            >
              Image
              <div
                class="uploader-input"
              >
                <input
                  class="uploader-content"
                  data-testid="uploader-img"
                  name="image"
                  type="file"
                />
              </div>
            </label>
          </div>
          <div
            class="group-btn"
          >
            <button
              class="button btn-secondary btn-small"
              type="button"
            >
              Create
            </button>
            <button
              class="button btn-warning btn-small"
              type="button"
            >
              Cancel
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <h2
        class="form-heading"
      >
        Create New Book
      </h2>
      <form
        class="book-form"
        data-testid="form"
      >
        <div
          class="input"
        >
          <label
            class="input-title"
          >
            Title
            <input
              class="input-content"
              name="title"
              type="text"
              value=""
            />
          </label>
        </div>
        <div
          class="input"
        >
          <label
            class="input-title"
          >
            Author
            <input
              class="input-content"
              name="author"
              type="text"
              value=""
            />
          </label>
        </div>
        <div
          class="input"
        >
          <label
            class="input-title"
          >
            Price
            <input
              class="input-content"
              name="price"
              type="text"
              value="0"
            />
          </label>
        </div>
        <div
          class="input"
        >
          <label
            class="input-title"
          >
            Description
            <textarea
              class="textarea-content"
              id="desc"
              name="desc"
              rows="3"
            />
          </label>
        </div>
        <div
          class="uploader"
        >
          <label
            class="uploader-title"
          >
            Image
            <div
              class="uploader-input"
            >
              <input
                class="uploader-content"
                data-testid="uploader-img"
                name="image"
                type="file"
              />
            </div>
          </label>
        </div>
        <div
          class="group-btn"
        >
          <button
            class="button btn-secondary btn-small"
            type="button"
          >
            Create
          </button>
          <button
            class="button btn-warning btn-small"
            type="button"
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
